<a href="https://colab.research.google.com/github/ShreyaGautamm/rnn-imdb-dataset/blob/main/example-implementation/Embedding_Example.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## One Hot Embedding Example

In [3]:
from tensorflow.keras.preprocessing.text import one_hot

In [2]:
###sentences
sent = [ 'the glass of milk',
        'the glass of juice',
         'the cup of tea',
         'I am a good boy',
         'I am a good developer',
         'understand the meaning of words',
         'your videos are good'
         ]

In [4]:
## Define the vocab size
voc_size = 10000

In [5]:
## One Hot Representation for every word
one_hot_repr = [one_hot(words, voc_size) for words in sent]
one_hot_repr


[[2583, 1610, 5203, 9900],
 [2583, 1610, 5203, 11],
 [2583, 776, 5203, 2781],
 [3295, 5097, 453, 8913, 7678],
 [3295, 5097, 453, 8913, 1254],
 [3825, 2583, 8867, 5203, 7954],
 [4972, 6454, 942, 8913]]

Every word is represented as its index in the vocabulary size. i.e., \
'the' is present at the 2583 index. \
'glass' is present at the 1610 index.

Now we will convert all these words to indexes and sentence by sentence pass it to embedding layer

In [7]:
## Word Embedding Repr

from tensorflow.keras.layers import Embedding
#from tensorflow.keras.processing.sequence import pad_sequences
from tensorflow.keras.utils import pad_sequences
from tensorflow.keras.models import Sequential

In [8]:
import numpy as np
import pandas as pd

We need to make all the sentences of fixed size for RNN, therefore, we use `pad_sequences`

In [9]:
sent_length = 8
embedded_docs = pad_sequences(one_hot_repr, padding='pre', maxlen=sent_length)
embedded_docs

array([[   0,    0,    0,    0, 2583, 1610, 5203, 9900],
       [   0,    0,    0,    0, 2583, 1610, 5203,   11],
       [   0,    0,    0,    0, 2583,  776, 5203, 2781],
       [   0,    0,    0, 3295, 5097,  453, 8913, 7678],
       [   0,    0,    0, 3295, 5097,  453, 8913, 1254],
       [   0,    0,    0, 3825, 2583, 8867, 5203, 7954],
       [   0,    0,    0,    0, 4972, 6454,  942, 8913]], dtype=int32)

In [10]:
## Feature Representation
dim=10 #10 feaures with respect to the dimensions in the word embeddings

In [14]:
model = Sequential()
model.add(Embedding(voc_size, dim, input_length=sent_length))
model.compile('adam', 'mse')

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


In [16]:
model.build(input_shape=(None, sent_length))
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)         │ (None, 8, 10)          │       100,000 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 100,000 (390.62 KB)

 Trainable params: 100,000 (390.62 KB)

 Non-trainable params: 0 (0.00 B)

In [17]:
model.predict(embedded_docs)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 248ms/step


array([[[-0.02361484, -0.04026965, -0.02742243,  0.02125107,
         -0.01570704, -0.03566948, -0.03930898, -0.01526004,
          0.00050055,  0.0193061 ],
        [-0.02361484, -0.04026965, -0.02742243,  0.02125107,
         -0.01570704, -0.03566948, -0.03930898, -0.01526004,
          0.00050055,  0.0193061 ],
        [-0.02361484, -0.04026965, -0.02742243,  0.02125107,
         -0.01570704, -0.03566948, -0.03930898, -0.01526004,
          0.00050055,  0.0193061 ],
        [-0.02361484, -0.04026965, -0.02742243,  0.02125107,
         -0.01570704, -0.03566948, -0.03930898, -0.01526004,
          0.00050055,  0.0193061 ],
        [-0.01702504, -0.00854455,  0.00565596,  0.02113884,
         -0.00799529,  0.04475418, -0.04195587, -0.02841766,
         -0.00012464, -0.04574344],
        [ 0.03304322,  0.04784426,  0.03914056, -0.03932036,
         -0.03854071, -0.00245813,  0.02417035,  0.00296646,
          0.01595979,  0.01101221],
        [ 0.00423706,  0.03602641, -0.02850316,  0.0